engine/notifications.js

```
@@ -0,0 +1,283 @@
+          return redis.sadd(`${deviceId}-notifications`, slug)
+          .then(() => redis.sadd(`${slug}-notifications`, deviceId));
+        }));
+      });
+    });
+  })
+  .then(() => getRegisteredFeatures(deviceId))
+  .catch(err => {
+    if (err.message !== 'Not Found') {
+      throw err;
+    }
+    return [];
+  })
+  .then(regFeatures => {
+    if (!doNotConfirm && endpoint) {
+      return sendConfirmation(endpoint, key, authSecret, deviceId, regFeatures)
+      .then(() => regFeatures);
+    }
+    return regFeatures;
+  });
+}
+
+// Registers to all notifications including new ones
+// Removes registrations to individual features
+function registerToAll(deviceId) {
+  return getRegisteredFeatures(deviceId)
+  .then(features => unregister(deviceId, features, true))
+  .then(() => Promise.all([
+    redis.sadd('all-notifications', deviceId),
+    redis.sadd(`${deviceId}-notifications`, 'all')]));
+}
+
+// registers to receive notifications
+// required info:
+// * deviceId
+// * list of feature slugs to register to
+//   special features are `new` and `all`
+//   * new - do notify about new features
+//   * all - all features including new, registering to all will remove
+//     all individual registrations
+// * endpoint [optional]
+// one can register to a feature without providing an endpoint
+// if it was already saved:
+// * key
+function register(deviceId, features, endpoint, key, authSecret) {
+  if (!features || features.length === 0) {
+    return Promise.reject(new Error('No features provided'));
+  }
+
+  return Promise.resolve()
+  .then(() => {
+    if (endpoint) {
+      const device = {
+        id: deviceId,
+        endpoint,
+        key: key || '',
+        authSecret: authSecret || '',
+      };
+      return redis.hmset(`device-${deviceId}`, device);
+    }
+  })
+  .then(() => redis.hgetall(`device-${deviceId}`))
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('No endpoint provided');
```

Can we throw this error at the beginning of the function?